message_endpoint.go

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright 2018 De-labtory
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package swim
+
+import (
+	"errors"
+
+	"github.com/DE-labtory/swim/pb"
+	"github.com/gogo/protobuf/proto"
+	"github.com/it-chain/iLogger"
+
+	"time"
+)
+
+var ErrSendTimeout = errors.New("Error send timeout")
+var ErrUnreachable = errors.New("Error this shouldn't reach")
+var ErrInvalidMessage = errors.New("Error invalid message")
+var ErrCallbackCollectIntervalNotSpecified = errors.New("Error callback collect interval should be specified")
+
+// sallback is called when target member sent back to local member a message
+// created field is for clean up the old callback
+type callback struct {
+	fn func(msg pb.Message)
+	created time.Time
+}
+
+// responseHandler manages callback functions
+type responseHandler struct {
+	callbacks map[string]callback
+	collectInterval time.Duration
+}
+
+func newResponseHandler(collectInterval time.Duration) *responseHandler {
+	h := &responseHandler{
+		callbacks: make(map[string]callback),
+		collectInterval: collectInterval,
+	}
+
+	go h.collectCallback()
+
+	return h
+}
+
+func (r *responseHandler) addCallback(seq string, cb callback) {
+	r.callbacks[seq] = cb
+}
+
+// Handle, each time other member sent back
+// a message, callback matching message's seq is called
+func (r *responseHandler) handle(msg pb.Message) {
+	seq := msg.Seq
+	cb, exist := r.callbacks[seq]
+
+	if exist == false {
+		iLogger.Panicf(nil, "Panic, no matching callback function")
+	}
+
+	cb.fn(msg)
+	delete(r.callbacks, seq)
+}
```

Comment:
@hihiboss 5 hours ago  Member
Does it need to go func??

  @zeroFruit
zeroFruit 3 hours ago  • 
 Member
I can't understand because in handle there's no go func

  @hihiboss
hihiboss 2 hours ago  Member
I meaned “Should the cbFn & delete func be executed as go func?”

  @zeroFruit
zeroFruit 28 minutes ago  • 
 Member
oh. you mean because resHandler.handle function blocking Listen function?